supply/lib/supply/uploader.rb

```diff
@@ -7,46 +7,115 @@ def perform_upload
-          upload_screenshots(language) unless Supply.config[:skip_upload_screenshots]
-          upload_changelogs(language) unless Supply.config[:skip_upload_metadata]
-        end
-      end
-
       apk_version_codes = []
       apk_version_codes.concat(upload_apks) unless Supply.config[:skip_upload_apk]
       apk_version_codes.concat(upload_bundles) unless Supply.config[:skip_upload_aab]
       upload_mapping(apk_version_codes)
 
       apk_version_codes.concat(Supply.config[:version_codes_to_retain]) if Supply.config[:version_codes_to_retain]
 
+
       # Only update tracks if we have version codes
       # Updating a track with empty version codes can completely clear out a track
       update_track(apk_version_codes) unless apk_version_codes.empty?
 
+
+      if !Supply.config[:rollout].nil? && Supply.config[:version_code].to_s != "" && Supply.config[:track].to_s != ""
+        update_rollout
+      end
+
       promote_track if Supply.config[:track_promote_to]
 
       if Supply.config[:validate_only]
-        UI.message("Validating all changes with Google Play...")
+        UI.message("Validating all track and release changes with Google Play...")
+        client.validate_current_edit!
+        UI.success("Successfully validated the upload to Google Play")
+      else
+        UI.message("Uploading all track and release changes to Google Play...")
+        client.commit_current_edit!
+        UI.success("Successfully finished the upload to Google Play")
+      end
+
+      perform_upload_meta(apk_version_codes)
+    end
+
+    def perform_upload_meta(version_codes)
+      client.begin_edit(package_name: Supply.config[:package_name])
+
+      if ((!Supply.config[:skip_upload_metadata] || !Supply.config[:skip_upload_images] || !Supply.config[:skip_upload_changelogs] || !Supply.config[:skip_upload_screenshots]) && metadata_path)
+        version_codes = [Supply.config[:version_code]] if version_codes.empty? 
+        version_codes.each do |version_code|
+          UI.user_error!("Could not find folder #{metadata_path}") unless File.directory?(metadata_path)
+
+          track, release = fetch_track_and_release(version_code)
+          UI.user_error!("Coult not find release for version code '#{version_code}' to update changelog") unless release
+
+          release_notes = []
+          all_languages.each do |language|
+            next if language.start_with?('.') # e.g. . or .. or hidden folders
+            UI.message("Preparing to upload for language '#{language}'...")
+
+            listing = client.listing_for_language(language)
+            
+            upload_metadata(language, listing) unless Supply.config[:skip_upload_metadata]
+            upload_images(language) unless Supply.config[:skip_upload_images]
+            upload_screenshots(language) unless Supply.config[:skip_upload_screenshots]
+            release_notes << upload_changelog(language, release.name) unless Supply.config[:skip_upload_changelogs]
```

Comment:
Still unsure if `skip_upload_changelogs` should just use `skip_upload_metadata`? 🤔

Reply:
I think we will keep it separate 👌